src/keychain_mac.cpp

```diff
@@ -101,67 +102,80 @@ void updateError(keychain::Error &err, OSStatus status) {
     }
 }
 
-void handleCFCreateFailure(keychain::Error &err,
-                           const std::string &errorMessage) {
+void setGenericError(keychain::Error &err, const std::string &errorMessage) {
+    err = keychain::Error{};
     err.message = errorMessage;
     err.type = keychain::ErrorType::GenericError;
     err.code = -1;
 }
 
-CFStringRef createCFStringWithCString(const std::string &str,
-                                      keychain::Error &err) {
-    CFStringRef result = CFStringCreateWithCString(
-        kCFAllocatorDefault, str.c_str(), kCFStringEncodingUTF8);
-    if (result == NULL) {
-        handleCFCreateFailure(err, "Failed to create CFString");
+/*! \brief Helper to manage the lifetime of CFObjects */
+template <typename T> struct ScopedCFRef {
+  public:
+    explicit ScopedCFRef(T &&ref) : _ref(std::move(ref)) {}
+    ~ScopedCFRef() {
+        if (_ref)
+            CFRelease(_ref);
```

Comment:
I would suggest to make this implementation a private member (`maybe_release()`) so that you can reuse it in the move-assignment constructor.

```suggestion
        if (_ref != nullptr) {
            CFRelease(_ref);
            _ref = nullptr;
        }
```

🐴 🤮